### ISSUE TITLE
Augmentations should specify allowed partial declarations.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#3690]
Augmentations can add missing pieces to existing declarations. If those missing pieces are *required*, then the initial declaration is not valid by itself.
The current specification would make that declaration a compile-time error.

Augmentations makes *some* such declarations still be compile-time errors that apply before augmentations are even applied, and other declarations be only partially declared, and filled in by an augmentation, in which case the property does not apply to the individual *declarations*, but to the entire declaration+augmentations stack, the "effective declaration".

*If* that allows a declaration to omit something that would otherwise be required by the *grammar*, we need to specify the allowed partial syntax too.

I'm not absolutely sure there are any cases where augmentations allow omitting something that is otherwise grammatically required, except maybe bodies of functions (which are grammatically allowed to be omitted, because of abstract declarations, but just not in every position - top-level functions cannot grammatically end in `;`).

It still means we need to go thrugh every "it's a compile-time error if declaration D ...." and decide whether that applies to the initial declaration syntax, or to the post-augmentation-application effective declaration. (And the latter case would imply that it's a semantic restriction, not a syntactic restriction.)

Some prose-introduced syntax errors are irrecoverable, which probably means being so invalid that applying an augmentation wouldn't make sense anyway. Other errors are not, and should be changed to only apply to the result of augmentation.

Examples:

* It's definitely and eagerly a compile-time error if a static method has a `covariant` parameter or a `this.x` or `super.x` parameter. We can't augment that, because it has no meaning. Anything that would add a meaning to the declaration that cannot possibly apply to that kind of declaration, should be an eager error.

* It's possibly not a compile-time error if a static method has a `;` body ("no body"), if a later applied augmentation adds a body.

* It's possibly not a compile-time error if a non-abstract class has an abstract member declaration, if a later augmentation adds a body (but that was never an eager error, because the class could inherit the implementation). Generally, anything related to satisfying an interface is going to be checked late.

* It's a compile-time error to have a `super.foo` invocation in an instance function, if the superclass doesn't have a `foo` member. It's probably allowed if an augmentation adds a superclass that does have a `foo`. (Which means the check should be written in a way that doesn't point to the surrounding class declaration, but to the semantic superclass of the surrounding class. It probably already is written that way, but we need to check. Everywhere, at least where we say "it's a compile-time error".)

**Comment by davidmorgan** [comment]
@johnniwinther this is the issue I'm aware of for the topic of "augmentations mean partial code can now be allowed".

We talked about something not in this issue, though, which is that the meaning of the partial code can be not yet known when macros ask about it. So for example

```
abstract class Foo {
  void doSomething();
}
```

if a macro asks about `doSomething` we have to choose whether we tell the macro it's abstract, not abstract, or that we don't know; and we have to choose what we do if the macro should ask again later; or otherwise, whether to inform it.

We discussed that indicating that you expect something to be augmented would resolve the ambiguity

```
abstract class Foo {
  void doSomething() augmented;
}
```

but that this does not look super attractive.

@lrhn @jakemac53 I couldn't find any other issues/discussion around this, do you know of any please?

It does seem like we need to work through the issue as per Lasse's opening description and all the consequences that follow, including for macro introspection.

**Comment by lrhn** [comment]
It's tricky. All we can safely say that the method does not appear to have a body.
If it's inside a non-abstract class or will _likely_ get a body, ... or the class has or will get a `noSuchMethod` implementation.

If we run macros in parallel, then two macros can both see a declaration with no body and generate a body that does not call `augmented()`.
The latter to be applied wins.
The fact that there is a "latter" suggests that macros are not truly parallel, and it would be better to let the latter macro see the results of the former. Even if we interleave phases, the macros of the individual phases should be serialized in a consistent way. 
(And then macros could generate new macro augmentations too.)

Or we can tell macro authors to always call `augmented` if they would if there was a body being augmented. (And we would allow calling if there is no augmented body, as a no-op.)

Are there any other properties that a macro can change significantly (other than mere existence in phase 1)?
Overriding fields with setters and getters doesn't change the signature, but it does change whether a final private instance variable is promotable.
Macro code shouldn't look at a private final instance variable and generate code that assumes it's promotable. (But code is generated on phase 3, and the augmenting getter would be added in phase 2?)

Adding elements to enums changes exhaustiveness, but again they're added in a phase before code containing switches.

It may just be adding bodies where no body existed before, that's changing something in phase 3.

**Comment by jakemac53** [comment]
This is why the current API uses `has*` instead of `is*` for these getters (for example `hasAbstract`) - it indicates only the presence of the keyword, and not concretely whether or not it is abstract.

For functions,  you also get `hasBody`, which tells you whether or not you can call `augmented()` safely.

**Comment by jakemac53** [comment]
> If we run macros in parallel, then two macros can both see a declaration with no body and generate a body that does not call `augmented()`.

No two macros which could generate a body for a function can run in parallel based on the ordering specification today.

**Comment by lrhn** [comment]
Using `external` is almost certainly wrong. It may work, but it's sending all the wrong signals, and even when it works, it's not better than `=> throw Unimplemented("Not augmented");` in any way.

The reason it works is that we are currently *very* lenient about `external` members, more than we should be, so that if a declaration is `external` and has no actual implementation supplied, it will just throw at runtime if called.

The intent of `external` should be that the declaration is concrete and that an implementation *is* provided by the compiler in some way. And we should make it a compile-time error if one isn't!
It shouldn't just be a "late" method where we dynamically check if it's defined or not. We don't need more dynamic checks at runtime.

**Comment by jakemac53** [comment]
> We build outlines before phase 3 so we need to have a way to tell whether a method is concrete without having to provide a body. Currently the only way to do this is to use `external` but that is wild west territory where only backends roam.

Hmm, this does sound fundamentally limiting. I think you might have to run phase 3.

Or, we will have to make macros explicitly signal ahead of time which methods they will be filling in.

**Comment by davidmorgan** [comment]
I discovered the other issue with discussion on this :) it's #3580 about adding an "augmented" modifier.

**Comment by davidmorgan** [comment]
I had a thought on this one :)

In cases where `hasBody` is a meaningful part of the declaration, should it be the case that macros that add a  body are required to do so in Phase 2? It's no longer _just_ a definition.

Maybe that helps the CFE, not sure whether it also comes with problematic downsides for macros.

**Comment by jakemac53** [comment]
Fwiw, I think that the CFE might want to run phase 3 regardless during outline creation, because there might be errors it is missing otherwise?

**Comment by jakemac53** [comment]
Lets re-focus this on the original issue (ignoring macros, this is just about augmentations). If we want to discuss macros further let's open a new issue (but my stance is, you cannot tell if a method is abstract until all 3 phases are ran).

Concretely, I believe the steps that need to be taken are:

- Allow top level and static method declarations to have no body (just a `;`). It is still an error if no augmentation provides a body though, but it isn't a parse error or eager error.
- Disallow augmentations of extension types to have no parenthesized representation type clause (if we have a better name for this please let me know). See https://github.com/dart-lang/language/pull/4003 which adds a TODO for this.
  - It might be simpler if the grammar just allows extension types to generally omit this, and then the error is reported separately based on whether it is an augmenting declaration of not.
- Disallow `on` clauses for extension and mixin augmentations.
  - This is already specified as an error, but we need to decide if it should be a parse error or not (similar to above, probably should make the same choice).  
- Anything else?

**Comment by munificent** [comment]
Maybe the right answer here is that we should make an official distinction between syntax errors and semantic errors?

* A syntax error is anything invalid based on the language grammar or prose that refines the grammar. Syntax errors are detectable and reportable before augmentations are applied. The formatter is not able to format code with syntax errors.

* A semantic error is anything else detected at compile time. Type errors, unimplemented abstract methods, inheritance chain cycles, resolution errors, etc. These can't be detected until after augmentations are applied.

We'd have to go through the spec and clarify for every compile-time error which kind it is. Probably a lot of tedious work. But, given that we *do* have a formatter that cares about this distinction, it wouldn't hurt to make it official. Up to this point, the formatter's rule about what it can format has been "whatever analyzer doesn't complain about as long as I don't try to resolve" which isn't *super* well-defined. It has occasionally caused problems where analyzer happens to only report an error during resolution that's pretty "syntactical" and the result is that the formatter ends up discarding some code and then triggering its own internal sanity check.

**Comment by lrhn** [comment]
> we should make an official distinction between syntax errors and semantic errors?

Yes. We should define when something is an error for a single syntactic declaration, and when it's an error for a full stack of declarations of the same entity.

Like, it's an error if a class declaration is both `abstract` and `sealed` (if that's not encoded on the grammar), but it's an error if a (fully augmented) non-abstract class doesn't implement its interface. 
Later augmentations can add both concrete members and new abstract members or entire super-interfaces.

We need to do that anyway, but looking at the existing errors in the specification, and decisive which bucket they go in, is a good start.

**Comment by munificent** [comment]
I see 434 occurrences of "compile-time error" in the language spec. I do think it's a good idea to make a distinction between syntax errors and other compile-time errors, but I'm not sure concretely how to go about that. The spec isn't fully up to date with recent language changes is it? Am I literally going to send a PR that replaces a selected set of "compile-time" in the spec LaTex file with "syntax"? Or will there be a giant section in the augmentations spec that lists every compile-time error in the spec that now becomes a syntax error?

I'm not asking facetiously. I want to make progress on this, I'm just not sure, like, *how*.

**Comment by munificent** [comment]
OK, I went through all 434 instances of "compile-time error" in the spec to try to see which ones might be relevant to augmentations. Obviously, every error could come from or in a declaration which happens to also use an augmentation. The cases that are interesting are ones where the introductory declaration either would or wouldn't have an error and an augmentation can flip that.

For all of those cases, my intention is that the behavior is what you get after the augmentation is applied. In other words, the compile errors should be the same as if you'd written all of the code in one monolithic declaration.

Here are the specified errors where I could an imagine an augmentation flipping the error state, examples that demonstrate that, and whether the resulting code should have an error or not:

### Collision with type parameter name

> If a generic class named G declares a type variable named X, it is a compile-time error if X is equal to G, or if G has a member whose basename is X, or if G has a constructor named G.X.

```dart
class C<T> {}

augment class C<T> {
  C.T();
}
```

Should be an error.

### Fully implementing an interface

> Let C be a concrete class C, I, m declared in library L, with interface I. Assume that I has a member signature m which is accessible to L. It is a compile-time error if C does not have a concrete member with the same name as m and accessible to L, unless C has a non-trivial noSuchMethod (10.2.2).

```dart
abstract class I {
  m();
}

class C implements I {}

augment class C {
  m() {}
}
```

Should be OK.

### Corresponding field for initializing formal

> A term of the form this.id is contained in p, and id is the name of p. It is a compile-time error if id is not also the name of an instance variable of the immediately enclosing class or enum.

```dart
class C {
  C(this.x);
}

augment class C {
  int x;
}
```

Should be OK.

### Corresponding field for constructor initializer

> An initializer of the form v = e is equivalent to an initializer of the form this.v = e, and both forms are called instance variable initializers. It is a compile-time error if the enclosing class does not declare an instance variable named v.

```dart
class C {
  C() : x = 1;
}

augment class C {
  int x;
}
```

Should be OK.

### Corresponding superclass constructor

> Let S be the superclass of the enclosing class of s. It is a compile-time error if class S does not declare a generative constructor named S (respectively S.id).

```dart
class C {}

class D extends C {
  D() : super.a(1);
}

augment class C {
  C.a(int x);
}
```

Should be OK.

### Uninitialized final field

> Let f be a final instance variable declared in the immediately enclosing f class or enum. A compile-time error occurs unless f is initialized by one of the following means:
>
> * f is declared by an initializing variable declaration.
> * f is initialized by means of an initializing formal of k.
> * f has an initializer in k’s initializer list.

```dart
class C1 {
  final int x;
}

augment class C1 {
  C1(this.x);
}

class C2 {
  final int x;
}

augment class C2 {
  C2() : x = 1;
}
```

Should be OK.

### Constructor initializer of unknown field

> It is a compile-time error if k’s initializer list contains an initializer for a variable that is not an instance variable declared in the immediately surrounding class.

```dart
class C {
  C() : x = 1;
}

augment class C {
  int x;
}
```

Should be OK.

### Const constructor on class with non-final field

> It is a compile-time error if a non-redirecting generative constant constructor is declared by a class that has a instance variable which is not final.

```dart
class C {
  const C();
}

augment class C {
  int x; // Not final.
}
```

Should be an error.

### Const constructor on class with non-const initializer

> If a non-redirecting generative constant constructor k is declared by a class C, it is a compile-time error for an instance variable declared in C to have an initializing expression that is not a constant expression.

```dart
class C {
  const C();
}

augment class C {
  final int x = DateTime.now().second;
}
```

Should be an error.

### Superclass cycle

> It is a compile-time error if a class C is a superclass of itself.

```dart
class C {}

augment class C extends C {}
```

Should be an error.

### Implementing the same type twice

> It is a compile-time error if two elements in the type list of the implements clause of a class C specifies the same type T.

```dart
class I {}

class C implements I {}

augment class C implements I {}
```

Should be an error.

### Extending and implementing the same type

> It is a compile-time error if the superclass of a class C is one of the elements of the type list of the implements clause of C.

```dart
class I {}

class C implements I {}
augment class C extends I {}
```

Should be an error.

### Implementing the same interface with different type arguments

> It is a compile-time error if a class C has two superinterfaces that are different instantiations of the same generic class.

```dart
class I<T> {}

class C implements I<int> {}
augment class C implements I<String> {}
```

Should be an error.

### Superinterface cycle

> It is a compile-time error if the interface of a class C is a superinterface of itself.

```dart
class A {}

class C implements A {}

augment class A implements C {}
```

Should be an error.

### Named constructor and static member collision

> Let C be a class. It is a compile-time error if C declares a constructor C named C.n and a static member with basename n.

```dart
class C {
  C.n();
}

augment class C {
  static n() {}
}
```

Should be an error.

### Static and instance member collision

> It is a compile-time error if C declares a static member with basename n and the interface of C has an instance member with basename n.

```dart
class C {
  static n() {}
}

augment class C {
  n() {}
}
```

Should be an error.

### Instance method and setter collision

> It is a compile-time error if the interface of C has an instance method named n and an instance setter with basename n.

```dart
class C {
  n() {}
}

augment class C {
  set n(_) {}
}
```

Should be an error.

### Static method and setter collision

> It is a compile-time error if C declares a static method named n and a static setter with basename n.

```dart
class C {
  static n() {}
}

augment class C {
  static set n(_) {}
}
```

Should be an error.

### Unknown superclass method

> Let Ssuper be the superclass (10.9) of the immediately enclosing class for i, Ssuper, L and let L be the library that contains i. Let the declaration D be the result D, F of looking up the method m in Ssuper with respect to L (17.18), and let F be the static type of D. Otherwise, if the method lookup failed, let the declaration D be the result of looking up the getter m with respect to L in Ssuper (17.18), and let F be the return type of D. If both lookups failed, a compile-time error occurs.

```dart
class A {
  a() {}
}

class C {
  m() { super.a(); }
}

augment class C extends A {}
```

Should be OK.

We'll definitely want to capture these in tests. But, beyond that, I'm not sure if any of this should be rolled into the proposal. Thoughts?

**Comment by lrhn** [comment]
The final specification should make the distinction between an error in a syntactic declaration and an error in the combined,/fully augmented/semantic declaration.

The proposal should probably just say that "all existing errors apply to the combined fully augmented declaration".

I'd actually consider whether these two should be on the individual syntactic declarations:

* Implementing the same type twice
* Extending and implementing the same type

The latter in particular, because we _should_ allow a `class C implements FancyEntity` to be implemented by a code-generated `augment class C extends FancyEntity;`.
(The alternative is to force the code generator to add `class _StupidLanguage$FancyEntity extends FancyEntity; class C extends _StupidLanguage$FancyEntity;`. Don't be that language! 😁) 

For the former, it's just not a problem if you write `implements Foo` twice on two separate syntactic declarations.
It's not *worth* making it a compile-time error, or forcing a code generator to check if the class already implements `Foo` before saying `implements Foo`. No harm, no foul.

These two rules only exist to catch a single user writing the same name twice in the same declaration.
With augmentations, there may be two authors, one being a code generator, and those rules don't fit with that.

(There wasn't any rule disallowing `with Foo implements Foo`? If there was, that should follow the `extends` behavior and be allowed if it's two different declarations. The `with` could be the code-generated implementation of the `implements`.)

**Comment by munificent** [comment]
> I'd actually consider whether these two should be on the individual syntactic declarations:
> 
> * Implementing the same type twice
> * Extending and implementing the same type
> 
> The latter in particular, because we _should_ allow a `class C implements FancyEntity` to be implemented by a code-generated `augment class C extends FancyEntity;`. (The alternative is to force the code generator to add `class _StupidLanguage$FancyEntity extends FancyEntity; class C extends _StupidLanguage$FancyEntity;`. Don't be that language! 😁)

These rules are weird because they are already inconsistent with mixins. This is an error:

```dart
class C extends I implements I {}
```

But this is not:

```dart
class C with M implements M {}
```

Filed https://github.com/dart-lang/language/issues/4542 to figure out what we want to do here.
